File: tutorials/004_configure_model_options/configure_model_options.cpp
// Model::Options chapter: configure input/preproc/boxdecode via Options, inspect specs.
//
// Usage:
//   tutorial_004_configure_model_options --model /path/to/yolo_v8s.tar.gz

#include "neat.h"

#include <opencv2/core.hpp>

#include <array>
#include <iostream>
#include <stdexcept>
#include <string>

namespace {

bool get_arg(int argc, char** argv, const std::string& key, std::string& out) {
  for (int i = 1; i + 1 < argc; ++i) {
    if (key == argv[i]) {
      out = argv[i + 1];
      return true;
    }
  }
  return false;
}

void print_spec(const char* label, const simaai::neat::TensorConstraint& spec) {
  std::cout << label << ": rank=" << spec.rank << " dtypes=" << spec.dtypes.size()
            << " shape_dims=" << spec.shape.size() << "\n";
}

} // namespace

int main(int argc, char** argv) {
  try {
    std::string model_path;
    if (!get_arg(argc, argv, "--model", model_path)) {
      std::cerr << "Usage: tutorial_004_configure_model_options --model <path>\n";
      return 1;
    }

    // Model::Options groups input caps, preproc, and box-decode into one struct.
    simaai::neat::Model::Options opt;
    // STEP set-input-preproc
    opt.preprocess.kind = simaai::neat::InputKind::Image;
    opt.preprocess.color_convert.input_format = simaai::neat::PreprocessColorFormat::BGR;
    opt.preprocess.input_max_width = 640;
    opt.preprocess.input_max_height = 640;
    opt.preprocess.input_max_depth = 3;
    opt.preprocess.normalize.enable = simaai::neat::AutoFlag::On;
    opt.preprocess.normalize.mean = std::array<float, 3>{0.485f, 0.456f, 0.406f};
    opt.preprocess.normalize.stddev = std::array<float, 3>{0.229f, 0.224f, 0.225f};
    // END STEP
    // STEP set-postproc
    opt.decode_type = simaai::neat::BoxDecodeType::YoloV8;
    opt.score_threshold = 0.55f;
    opt.nms_iou_threshold = 0.45f;
    opt.top_k = 100;
    opt.boxdecode_original_width = 640;
    opt.boxdecode_original_height = 640;
    opt.name_suffix = "_chapter";
    // END STEP

    // CORE LOGIC
    // STEP load-and-inspect
    simaai::neat::Model model(model_path, opt);
    print_spec("input_spec", model.input_spec());
    print_spec("output_spec", model.output_spec());
    std::cout << "metadata_keys=" << model.metadata().size() << "\n";
    // END STEP
    // END CORE LOGIC

    // STEP run-inference
    cv::Mat bgr(640, 640, CV_8UC3, cv::Scalar(10, 20, 30));
    if (!bgr.isContinuous())
      bgr = bgr.clone();
    auto out = model.run(std::vector<cv::Mat>{bgr}, /*timeout_ms=*/2000);
    std::cout << "outputs=" << out.size() << "\n";
    // END STEP
    std::cout << "[OK] 004_configure_model_options\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/004_configure_model_options/configure_model_options.py
#!/usr/bin/env python3
"""Configure a YOLO Model with ModelOptions knobs and introspect its specs.

Usage:
  python3 configure_model_options.py --model /path/to/yolo_v8s.tar.gz
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

try:
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either Neat is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate"
  )

import numpy as np


def main(argv: list[str]) -> int:
  ap = argparse.ArgumentParser(description=__doc__)
  ap.add_argument("--model", type=Path, required=True)
  args = ap.parse_args(argv[1:])

  opt = pyneat.ModelOptions()
  # STEP set-input-preproc
  opt.preprocess.kind = pyneat.InputKind.Image
  opt.preprocess.color_convert.input_format = pyneat.PreprocessColorFormat.BGR
  opt.preprocess.input_max_width = 640
  opt.preprocess.input_max_height = 640
  opt.preprocess.input_max_depth = 3
  opt.preprocess.normalize.enable = pyneat.AutoFlag.On
  opt.preprocess.normalize.mean = [0.485, 0.456, 0.406]
  opt.preprocess.normalize.stddev = [0.229, 0.224, 0.225]
  # END STEP
  # STEP set-postproc
  opt.decode_type = pyneat.BoxDecodeType.YoloV8
  opt.score_threshold = 0.55
  opt.nms_iou_threshold = 0.45
  opt.top_k = 100
  opt.boxdecode_original_width = 640
  opt.boxdecode_original_height = 640
  opt.name_suffix = "_chapter"
  # END STEP

  # CORE LOGIC
  # STEP load-and-inspect
  model = pyneat.Model(str(args.model), opt)
  print(f"input_shape={model.input_spec().shape}")
  print(f"output_shape={model.output_spec().shape}")
  print(f"metadata_keys={len(model.metadata())}")
  # END STEP
  # END CORE LOGIC

  # STEP run-inference
  bgr = np.full((640, 640, 3), 44, dtype=np.uint8)
  tensor = pyneat.Tensor.from_numpy(bgr, copy=True, image_format=pyneat.PixelFormat.BGR)
  outputs = model.run([tensor], timeout_ms=2000)
  print(f"output_count={len(outputs)}")
  # END STEP
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))
